overlord/snapstate/snapmgr.go

```diff
@@ -797,6 +798,12 @@ func (m *SnapManager) StartUp() error {
 	// the set of inhibited snaps is changed.
 	m.changeCallbackID = m.state.AddChangeStatusChangedHandler(processInhibitedAutoRefresh)
 
+	if AssertRuntimeFailureRestart(m.state) == ErrUnexpectedRuntimeFailure {
+		logger.Noticef("detected a runtime failure without a corresponding snapd change")
+		// restart manager is added before snap manager
+		restart.Request(m.state, restart.StopDaemon, nil)
```

Comment:
does this work done so early?

Reply:
this needs a test in managers_test.go perhaps?

Reply:
Yes, restart manager is set up early in overlord.New(). I'll add a managers level test.

The problem with this approach is that the state engine still tries to run, so that it may try to pick up any outstanding tasks, and the overlord loop is essentially started. Then the daemon tries to expose the API, although AFAICT the tomb it uses is dead on arrival. Still since overlord's loop gets run there's a slight chance that something eg. an auto refresh may be attempted. I'm working on a tweak for this.

Reply:
yes, I thought the intention was to try not to get too far into actually runnning things, a sepcial error and some extra code might be better than risking running things we don't want to run

Reply:
I have pushed a couple of changes so that snapd aborts earlier. Please have a look if the direction looks good.

Reply:
new approach looks better